### PR TITLE
WIP: analytics client and server

### PR DESCRIPTION
**Pull request by olivierteboul**
Analytics computations are heavy. Hence this PR introduces:
- a separate server to make sure those run in a different process
- the periodic callback is run there.
- an abstraction of dataset that can cache heavy preprocessing with some ttl.
- a client to this server: if some dataset is asked from another server (backoffice or www), we fetch the data from the analytics server in order to always have access to the cache and not run analytics computations outside of the analytics process.

### CODE REVIEW

File: icubam/analytics/dataset.py [hunk 1]
```diff
@@ -25,8 +30,78 @@
...
+      )
+    else:
+      result = self.db.get_bed_counts(max_date=max_ts)
+
+    result = store.to_pandas(result, max_depth=2)
+    result = result.drop(
+      columns=['icu_bed_counts', 'icu_users', 'icu_managers']
+    )
+    if preprocess:
+      result = preprocessing.preprocess_bedcounts(result)
```
Comment (rth) [inline icubam/analytics/dataset.py:92]: I think `preprocessing=True` and `latest=True` should raise an error.

**Comment by olivierteboul** [comment]
@rth thanks for the review. 
To answer your main comment, the problem is that:
- we want the heavy stuff to run in a separate process
- we want to have a cache.

So either the cache is like a redis cache (i am totally fine with it). Which could be a good idea. Or we cache in memory, and then it has to be in the memory of the heavy process. I have implemented it this way (but i would be happy to use redis). And therefore there is a slight indirection as you noticed. The call does not go directly to the analytics server (to avoid replicating the access key logics), but once passed the access key control, it ask the server with a cache to deliver the data.

Let me know what you think about it. We don't have to merge this solution if it is not considered good, but i believe we could further use the analytics server (and client) to get the json curves from it....

**Comment by rth** [comment]
Thanks for the explanations @olivierteboul , I think the current caching approach is good and redis would make the installation more difficult for other potential contributors which is probably not worth it in our use case.

My questions is more can't we make e.g. `/db/all_bedcounts` hit directly the analytics server (in production) instead of the www server and proxying it to the analytics server by tornado. But that would complexify somewhat the NGNIX routing setup I imagine?  Anyway not a blocker to merge this, just something I was wondering about.

**Comment by dulacarnold** [comment]
> Thanks for the explanations @olivierteboul , I think the current caching approach is good and redis would make the installation more difficult for other potential contributors which is probably not worth it in our use case.
> 
> My questions is more can't we make e.g. `/db/all_bedcounts` hit directly the analytics server (in production) instead of the www server and proxying it to the analytics server by tornado. But that would complexify somewhat the NGNIX routing setup I imagine? Anyway not a blocker to merge this, just something I was wondering about.

It might make sense to just redirect /db/ to the analytics server.  Currently if we ask for the full dump w/ preprocessing it blocks the prod server from displaying the map.... which is not great ;)

File: icubam/analytics/dataset.py [hunk 2]
```diff
@@ -25,8 +30,78 @@
...
+class Dataset:
+  """A class to manipulate the bedcounts data. With caching support."""
+  COLLECTIONS = ['icus', 'regions', 'bedcounts', 'all_bedcounts']
+
+  def __init__(self, db, ttl: int = 0):
+    self.db = db
+    self.ttl = ttl
+
+  @cached
+  def get_bedcounts(self, max_ts=None, latest=False, preprocess=True):
```
[the remark this answers is not in the document]
Comment (dulacarnold) [reply icubam/analytics/dataset.py:75]: I was thinking cached=false or something.

[the remark this answers is not in the document]
Comment (rth) [reply icubam/analytics/dataset.py:75]: That could also work. It's maybe just a bit more complex with the decorator.

**Pull request merged**